File: libs/shared-auth/src/main/java/br/com/deltaglobalbank/sharedauth/JwtValidator.java
package br.com.deltaglobalbank.sharedauth;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.List;
import java.util.UUID;

import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.jwk.source.JWKSource;
import com.nimbusds.jose.jwk.source.JWKSourceBuilder;
import com.nimbusds.jose.proc.JWSVerificationKeySelector;
import com.nimbusds.jose.proc.SecurityContext;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.proc.DefaultJWTProcessor;

public class JwtValidator {

    private final JwtValidationProperties properties;
    private final JWSVerificationKeySelector<SecurityContext> keySelector;

    public JwtValidator(JwtValidationProperties properties) {
        this.properties = properties;
        JWKSource<SecurityContext> jwkSource =
            JWKSourceBuilder.<SecurityContext>create(toUrl(properties.jwksUrl())).build();
        this.keySelector = new JWSVerificationKeySelector<>(JWSAlgorithm.RS256, jwkSource);
    }

    public ValidatedJwt validate(String token) {
        DefaultJWTProcessor<SecurityContext> processor = new DefaultJWTProcessor<>();
        processor.setJWSKeySelector(keySelector);

        JWTClaimsSet claims;
        try {
            claims = processor.process(token, null);
        } catch (Exception ex) {
            throw new JwtValidationException("Token inválido: " + ex.getMessage(), ex);
        }

        if (!properties.expectedIssuer().isEmpty()
            && !properties.expectedIssuer().equals(claims.getIssuer())) {
            throw new JwtValidationException("Issuer inválido: " + claims.getIssuer());
        }

        if (!properties.expectedAudience().isEmpty()
            && !claims.getAudience().contains(properties.expectedAudience())) {
            throw new JwtValidationException("Audience inválida");
        }

        return extractValidatedJwt(claims);
    }

    private ValidatedJwt extractValidatedJwt(JWTClaimsSet claims) {
        try {
            List<String> roles = claims.getStringListClaim("roles");
            List<String> modules = claims.getStringListClaim("modules");
            Object mustChange = claims.getClaim("must_change_password");
            return new ValidatedJwt(
                UUID.fromString(claims.getJWTID()),
                UUID.fromString(claims.getSubject()),
                UUID.fromString(claims.getStringClaim("tenant_id")),
                claims.getStringClaim("principal_type"),
                roles != null ? roles : List.of(),
                modules != null ? modules : List.of(),
                mustChange instanceof Boolean b ? b : false,
                claims.getExpirationTime().toInstant(),
                claims.getIssueTime().toInstant()
            );
        } catch (java.text.ParseException ex) {
            throw new JwtValidationException("Claims inválidas: " + ex.getMessage(), ex);
        }
    }

    private static URL toUrl(String value) {
        try {
            return new URL(value);
        } catch (MalformedURLException ex) {
            throw new IllegalArgumentException("jwksUrl inválida: " + value, ex);
        }
    }
}

File: libs/shared-auth/src/main/kotlin/br/com/deltaglobalbank/sharedauth/JwtValidator.kt
package br.com.deltaglobalbank.sharedauth

import com.nimbusds.jose.JWSAlgorithm
import com.nimbusds.jose.jwk.source.JWKSourceBuilder
import com.nimbusds.jose.proc.JWSVerificationKeySelector
import com.nimbusds.jose.proc.SecurityContext
import com.nimbusds.jwt.JWTClaimsSet
import com.nimbusds.jwt.proc.DefaultJWTProcessor
import java.net.URL
import java.time.Instant
import java.util.UUID

class JwtValidator(private val properties: JwtValidationProperties) {

    private val jwkSource = JWKSourceBuilder.create<SecurityContext>(URL(properties.jwksUrl)).build()
    private val keySelector = JWSVerificationKeySelector<SecurityContext>(JWSAlgorithm.RS256, jwkSource)

    fun validate(token: String): ValidatedJwt {
        val processor = DefaultJWTProcessor<SecurityContext>().apply {
            jwsKeySelector = keySelector
        }

        val claims = try {
            processor.process(token, null)
        } catch (ex: Exception) {
            throw JwtValidationException("Token inválido: ${ex.message}", ex)
        }

        if (properties.expectedIssuer.isNotEmpty() && claims.issuer != properties.expectedIssuer) {
            throw JwtValidationException("Issuer inválido: ${claims.issuer}")
        }

        if (properties.expectedAudience.isNotEmpty() && !claims.audience.contains(properties.expectedAudience)) {
            throw JwtValidationException("Audience inválida")
        }

        return extractValidatedJwt(claims)
    }

    private fun extractValidatedJwt(claims: JWTClaimsSet): ValidatedJwt {
        return ValidatedJwt(
            jti = UUID.fromString(claims.jwtid),
            subject = UUID.fromString(claims.subject),
            tenantId = UUID.fromString(claims.getStringClaim("tenant_id")),
            principalType = claims.getStringClaim("principal_type"),
            roles = claims.getStringListClaim("roles") ?: emptyList(),
            modules = claims.getStringListClaim("modules") ?: emptyList(),
            mustChangePassword = (claims.getClaim("must_change_password") as? Boolean) ?: false,
            expiresAt = claims.expirationTime.toInstant(),
            issuedAt = claims.issueTime.toInstant()
        )
    }
}
